### PR TITLE
Added find feature to Select option in Find dialog

### LINKED ISSUE #8217: Add a Find Columns dialogue into the Edit menu
@N-thony you are working on the Find dialogue now.  There is becoming more work, so you may wish to allocate or share this work.

I suggest what you are working on becomes `Find Rows`, instead of just Find.  We don't need Find Next in the menu and I suggest we replace it by `Find Columns`.  This will again have 2 options and the select is the second - just like Filter is the second in the Find rows.  And it will indicate a column, just as the Find Rows indicates a row.

I strongly suggest the Select option be done first, because the ordinary option will take longer, and will essentially be a copy of the Select code.  It will then jump to the correct "page" sideways, just are your current find jumps to the right page going downwards.

### PR DESCRIPTION
**Pull request by N-thony**
Fixes #8217
@rdstern I have now implemented the find feature in Select option, have a look. Thanks.

### CODE REVIEW

**Review by rdstern — COMMENTED** [review]
@N-thony this is great and almost seems to work already!
I tried with the hh.sav file I have sent you - in case you didn't have it.
I made a selection of about 7 variables (doesn't matter which, but I scattered them and then had some close together.: HH2, HH5, 33X.34, 33X.36, 33X.37, ST4N.1 and 2 and 3).

a) I don't think it found them all correctly - and I occasionally clicked multiple times and it seemed to stay where it was - maybe it had made later variables greeen, but they were not visible to me?
b) When I closed the dialog a green column remained.  I am happy they remain, but how should we reset them, when we don't want them green?
c) I am wondering whether you add the usual buttons at the bottom?  That's for all 3 cases.  If so, then Close is just the same as the usual close.  Could Reset them be used to get rid of the marked rows or columns as well as resetting within the dialog.  And Help will be useful anyway.  I need to get the number.  I presume Ok and script remain disabled - is there R-code for script?  
d) One item missing on the columns is which column you have found.  For the rows in your ordinary Find I note that you have a message when searching for someting that isn't there.  Could there also be the opposite - and possibly for all 3 options, but particularly useful for the select.  It would give the column number as well as the name.  (You know the number when looking at the metadata.  )
You have space either above or below the Find.  For columns It could be 
Found Variable: xxx  
Name: yyy
Label: zzz

**Review by rdstern — COMMENTED** [review]
@N-thony that is looking pretty good.
a) I was hoping that the Found Variable : 129, so would give the number.  Currently you get the name twice.
b) I am not so keen on the dialog expanding to the length of the longest label.  Could the size stay fixed.  If necessary I can scroll or have the label on multiple lines here?
c) I am not so sure about the green.  I missed it at the start.

**Review by rdstern — APPROVED** [review]
@N-thony this is great.  It is easy to think of improvements, but let's get this merged first - as soon as @lloyddewit is also happy.

File: instat/dlgFindInVariableOrFilter.vb [hunk 1]
```diff
@@ -118,78 +132,165 @@ Public Class dlgFindInVariableOrFilter
...
+                Dim iColumn As Integer = GetColumnIndex(strColumn)
+                Dim iColPage As Integer = Math.Ceiling(CDbl(iColumn / frmMain.clsInstatOptions.iMaxCols))
+                frmMain.ucrDataViewer.GoToSpecificColumnPage(iColPage)
+                frmMain.ucrDataViewer.SearchColumnInGrid(strColumn)
+
+                lblVariableFound.Text = "Found Variable: " & GetColumnIndex(strColumn) + 1
+                Dim strName = "Name: " & strColumn
+                TruncateLabelText(lblName, strName, 135)
+                Dim strLabel = "Label: " & GetColLabel(strColumn)
+                TruncateLabelText(lblLabel, strLabel, 135)
```
Comment (lloyddewit) [inline instat/dlgFindInVariableOrFilter.vb:227]: See my comment above about making TruncateLabelText a 'pure' function.
```suggestion
                lblName.Text = TruncateLabelText(lblName, strName, 135)
                Dim strLabel = "Label: " & GetColLabel(strColumn)
                lblLabel.Text = TruncateLabelText(lblLabel, strLabel, 135)
```

File: instat/dlgFindInVariableOrFilter.vb [hunk 2]
```diff
@@ -118,78 +132,165 @@ Public Class dlgFindInVariableOrFilter
 
     Private Sub TestOkEnabled()
         cmdFind.Enabled = (rdoVariable.Checked AndAlso Not ucrReceiverVariable.IsEmpty AndAlso Not ucrInputPattern.IsEmpty) OrElse
-            (rdoInFilter.Checked AndAlso Not ucrReceiverVariable.IsEmpty)
+            ((rdoInFilter.Checked OrElse rdoSelect.Checked) AndAlso Not ucrReceiverVariable.IsEmpty)
+    End Sub
+
+    Private Sub TruncateLabelText(label As Label, strName As String, maximumWidth As Integer)
```
Comment (lloyddewit) [inline instat/dlgFindInVariableOrFilter.vb:138]: Would it be cleaner to make this a 'pure' function (i.e. no side effects on the `label` parameter)?
```suggestion
    Private Function TruncateLabelText(label As Label, strName As String, maximumWidth As Integer) As String
```

File: instat/dlgFindInVariableOrFilter.vb [hunk 3]
```diff
@@ -118,78 +132,165 @@ Public Class dlgFindInVariableOrFilter
...
+            Dim truncatedWidth As Integer = originalWidth
+
+            While truncatedWidth > maximumWidth AndAlso truncatedText.Length > 0
+                truncatedText = truncatedText.Substring(0, truncatedText.Length - 1)
+                truncatedWidth = CInt(graphics.MeasureString(truncatedText & "...", font).Width)
+            End While
+
+            label.Text = truncatedText & "..."
+        Else
+            label.Text = strName
```
Comment (lloyddewit) [inline instat/dlgFindInVariableOrFilter.vb:154]:
```suggestion
            Return truncatedText & "..."
        Else
            Return strName
```

File: instat/dlgFindInVariableOrFilter.vb [hunk 4]
```diff
@@ -118,78 +132,165 @@ Public Class dlgFindInVariableOrFilter
...
+        End If
     End Sub
 
+
```
Comment (lloyddewit) [inline instat/dlgFindInVariableOrFilter.vb:158]:
```suggestion
```

File: instat/dlgFindInVariableOrFilter.vb [hunk 5]
```diff
@@ -118,78 +132,165 @@ Public Class dlgFindInVariableOrFilter
 
     Private Sub TestOkEnabled()
         cmdFind.Enabled = (rdoVariable.Checked AndAlso Not ucrReceiverVariable.IsEmpty AndAlso Not ucrInputPattern.IsEmpty) OrElse
-            (rdoInFilter.Checked AndAlso Not ucrReceiverVariable.IsEmpty)
+            ((rdoInFilter.Checked OrElse rdoSelect.Checked) AndAlso Not ucrReceiverVariable.IsEmpty)
+    End Sub
+
+    Private Sub TruncateLabelText(label As Label, strName As String, maximumWidth As Integer)
+        Dim graphics As Graphics = label.CreateGraphics()
+        Dim originalText As String = strName
```
Comment (lloyddewit) [inline instat/dlgFindInVariableOrFilter.vb:140]: do we need this?

File: instat/dlgFindInVariableOrFilter.vb [hunk 6]
```diff
@@ -118,78 +132,165 @@ Public Class dlgFindInVariableOrFilter
...
+        expItems = frmMain.clsRLink.RunInternalScriptGetValue(clsColmnLabelsRFunction.ToScript(), bSilent:=True)
+
+        If expItems IsNot Nothing AndAlso Not (expItems.Type = Internals.SymbolicExpressionType.Null) Then
+            Dim strArr As String() = expItems.AsCharacter.ToArray
+            If strArr IsNot Nothing Then
+                'the number of labels for a column expected is 1
+                If strArr.Length = 1 Then
+                    strColLabel = strArr(0)
+                ElseIf strArr.Length > 1 Then
+                    MessageBox.Show(Me, "Developer error: retrieved column label should be one.", "Developer Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
```
Comment (lloyddewit) [inline instat/dlgFindInVariableOrFilter.vb:272]: make message more meaningful?
```suggestion
                    MessageBox.Show(Me, "Developer error: more than one column label found.", "Developer Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
```

File: instat/dlgFindInVariableOrFilter.vb [hunk 7]
```diff
@@ -20,9 +20,10 @@ Public Class dlgFindInVariableOrFilter
     Private bFirstLoad As Boolean = True
     Private bReset As Boolean = True
     Private iCurrentOccurenceIndex As Integer
-    Private iCountClick As Integer
+    Private iCountRowClick, iColumnClick As Integer
```
Comment (lloyddewit) [inline instat/dlgFindInVariableOrFilter.vb:23]: is it safer yo initialise these to `1` (either here or in `SetDefaults()` or `initialiseDialog()`?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@N-thony This was a significant piecw of work, it looks good. Just a few comments, thanks

File: instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb [hunk 8]
```diff
@@ -282,8 +283,13 @@ Public Class ucrDataViewReoGrid
         Return lstRowsIndexes
     End Function
 
-    Private Sub ScrollToCellPos(currWorkSheet As Worksheet, iRow As Integer, iCol As Integer)
-        currWorkSheet.FocusPos = currWorkSheet.Cells(row:=iRow, col:=iCol).Position
+    Private Sub ScrollToCellPos(currWorkSheet As Worksheet, iRow As Integer, iCol As Integer, bApplyToRow As Boolean)
+
+        If bApplyToRow Then
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb]: @N-thony I'm still a bit confused. Please could you clarify if this is correct?

**Comment by lloyddewit** [comment]
@N-thony Thank you for the changes. There is one open comment above, please could you check it?

File: instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb (hunk 8 above)
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb]: OK, I think it makes sense, thanks

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern there have been some changes since your last approval. If you can retest/approve, then we can merge, thanks

**Review by rdstern — APPROVED** [review]
Great.  Many thanks to both of you - this is an important new feature!

**Pull request merged**